src/authentication/background/auth-service.ts

```diff
@@ -0,0 +1,121 @@
+import {
+    AuthEvents,
+    AuthInterface,
+    AuthRemoteFunctionsInterface,
+} from 'src/authentication/background/types'
+import {
+    Claims,
+    UserPlans,
+    UserFeatures,
+} from 'firebase-backend/firebase/functions/src/types'
+
+import { RemoteEventEmitter } from 'src/util/webextensionRPC'
+
+export class AuthSubscriptionError extends Error {}
+export class AuthSubscriptionInvalid extends AuthSubscriptionError {}
+export class AuthSubscriptionNotPresent extends AuthSubscriptionInvalid {}
+export class AuthSubscriptionExpired extends AuthSubscriptionInvalid {}
+
+//
+/**
+ * Number of milliseconds leeway after a subscription has expired
+ * Should be at least as long as the auth implementation's refresh token
+ * so for firebase, 1h.
+ */
+const subscriptionGraceMs = 1000 * 60 * 60
+
+export class AuthService implements AuthRemoteFunctionsInterface {
+    private readonly auth: AuthInterface
+
+    public constructor(authImplementation: AuthInterface) {
+        this.auth = authImplementation
+    }
+
+    // TODO: (ch) (opt-out-auth): to allow the user to enable/disabled auth we also need to re-initialise the background auth service
+    // static get isEnabledByUser() {
+    //     return localStorage.getItem('AUTH_ENABLED') === 'true'
+    // }
+    // static setEnabledByUser() {
+    //     localStorage.setItem('AUTH_ENABLED', 'true')
+    // }
+
+    public registerAuthEmitter(emitter: RemoteEventEmitter<AuthEvents>) {
+        this.auth.registerAuthEmitter(emitter)
+    }
+
+    public getUser = async () => this.auth.getCurrentUser()
+    private getUserClaims = async () => this.auth.getUserClaims()
+    public refresh = async () => this.auth.refresh()
+
+    /**
+     *  Checks that a client has a valid subscription (exists, is not expired)
+     *  to the provided plan.
+     */
+    public checkValidPlan = async (plan: UserPlans): Promise<boolean> => {
+        const claims = await this.getUserClaims()
+
+        const subscriptionExpiry = this.subscriptionExpiryAccessor(claims)(plan)
+
+        if (!subscriptionExpiry) {
+            throw new AuthSubscriptionNotPresent()
+        }
+
+        if (
+            new Date().getUTCMilliseconds() >=
+            subscriptionExpiry + subscriptionGraceMs
+        ) {
+            throw new AuthSubscriptionExpired()
+        }
+
+        return true
+    }
+
+    private subscriptionExpiryAccessor = (claims: Claims) => (
```

Comment:
Even if it's a private method, it'd be better to always use more descriptive names, even if they seem a bit redundant. Here it could be `getSubscriptionExpirationTimestamp()`. There's also the `withClaims()` in another file.